### ISSUE TITLE
GitHub issues are too hard to navigate and there are too many of them

### ISSUE THREAD

**Issue by brianlheim** [supercollider/supercollider#4940]
## Motivation

the motivation here is **to make the issue tracker work better for maintainers**.

i just realized that as a maintainer, i have reached a point where my relationship to the issues list on GitHub is truly awful. i barely ever look at it, and when i do, it's because i want to keep the count below 600 and close some pointless tickets or solve some easy bugs. i just keep a list of the things i want to work on in my head or on my computer. i avoid logging new issues at all because i see creating the issue ticket as a waste of time unless i know someone else or future me is going to take it on.

having a high number of issues open with no way to navigate them is psychically and spiritually fatiguing for me. i would guess it is that way for others given the number of failed attempts we've had at organizing and maintaining project boards and milestones.

the issues list is supposed to facilitate communication and coordination between users and developers of the project, and among developers. right now, it does not serve that role well, and often gets in the way.

## Description of Proposed Feature

this is a fairly open ended ticket and i would love to get others' thoughts.

here's how i'm thinking of solving this:

1. add labels to aid people who are interested in helping.
2. remove labels that are unclear, unhelpful, or redundant.
3. add a clear reference for use of labels.
4. do not use milestones
5. decide on guidelines for when an issue can be closed.
6. be more aggressive about closing issues.

### labels

i would suggest 3 sets of labels:
- severity: how serious the problem is; rough indicator of how seriously people are impacted by this
- difficulty: how much effort is required to investigate and solve the problem; a rough and relative estimate since this depends on familiarity with the language, codebase, and section of the codebase, among other things.
- blocking: indicates whether the issue can be tackled immediately with investigation and a PR, or whether it requires another issue, research, discussion, or even an RFC

severity would be required on every issue, and one or both of difficulty and blocking, since sometimes the difficulty of a solution won't be known until there is more info or discussion to guide a solution.

i've seen these categories used in other issue trackers like Bugzilla and JIRA, so i think they are at least reasonable ideas. i would definitely find them helpful.

if i were going to tag this issue according to these categories, i would choose "severity: high" (not "critical"), "difficulty: medium" (because of the effort required for both discussion and implementation), and "blocked by: discussion".

### milestones

i don't want to use them, at least not right now. we tried before, and i do give @snappizz credit for trying to make it work. but, i think they turned out to be a lot of maintenance effort, were never quite clear enough in practice, and are kind of replaced by the "3.y.z discussions" tickets. Nathan, i'd love to hear any thoughts you had on milestones even if you aren't doing maintenance work at the moment.

### closing issues

right now we close a ticket for the following reasons, which arose informally and somewhat experimentally:
- spam
- wrong project / not our problem (move to right project if possible)
- duplicate
- intended behavior
- not enough information
- "stale"
- a vague concept of out-of-scope / too difficult / etc.

the first 5 are clear, but i would like to clear up the last 2, add "unsupported platform", and document these as a reference for maintainers and others.

here's what i think:

"stale" means we were waiting on information or discussion for more than 2 weeks. after 2 weeks of silence, it is likely in my experience that no conclusion will ever be reached, and it is better to close. it is also a recognition that there is a natural limit to the number of ongoing discussions that can be happening at any given time on this project, and that trying to split our attention among all those discussions can often lead to less work getting done overall. however, this 2-week rule shouldn't be too rigid -- for instance, issues with high severity should be kept open even if discussion is stalled for a bit, or perhaps they should be bumped up to "RFC required" if the reason it stalled was a lack of consensus or research.

"out of scope" means that the issue would require adding functionality that we don't want to be responsible for, or fixing something that should be deprecated and removed. this is of course difficult to get consensus on, but i think we should be proactive (*and respectful*) about speaking up when we feel that a new feature would be out of scope. perhaps over time we can arrive at some shared ideas of what exactly the "scope" of SuperCollider is exactly.

"too difficult" is a general way of saying, this issue would take so much effort and provide so little value or help so few people that we do not plan to solve it in this century. it's a function of both severity and difficulty.

we could also consider closing issues when the underlying problem is already solved conveniently enough by some other tool. "workaround exists". this would be a function of difficulty and effort as well.

## Plan for Implementation

i would happily volunteer to be a person to go through all open issues and re-label them. that's how much i believe this would help. probably having a few people do it would be good so we can create a shared understanding of how we use the new labels. i am also happy to write the documentation for this if others can review it, or vice versa.

**Comment by brianlheim** [comment]
FYI -- happy to have suggestions/thoughts even if you haven't taken the time to read everything above, i know it's a lot. the core of it is the numbered list under "Description of proposed feature".

**Comment by jrsurge** [comment]
In GitHub Actions, there's an automation workflow that might help with managing, at least some of, this:

https://github.com/actions/stale

The configuration options for it can be seen here:
https://github.com/actions/stale/blob/master/action.yml

**Comment by brianlheim** [comment]
yes great! if we're making this a little more formal we should definitely try adding automation.

another idea i just had -- maybe we could set a project cap on the number of issues open at a time, and once we go over that cap start closing issues in order of oldest / least severe? not sure if that is too draconian, but it would make me feel much much better, and i think it's fair to say that even at 400 or 500 issues there is still plenty of work for everyone. :)

**Comment by geoffroymontel** [comment]
I'm OK with the proposition. The difficulty tag is difficult to assess because it depends on the skills of the person who tackles the bug. I agree that the milestone is not needed now that there are Github Projects and that an issue can be linked to a project.

**Comment by telephon** [comment]
Thank you for this, it'll be helpful to be able to manage the issues better.

Things to consider with "staling" / closing issues:
- How to distinguish issues that are closed because they are solved and those closed while still unsolved?
- How to encourage reopening of unsolved stale issues?
- How to remind people of the fact that there is an answer pending?

I notice that all these have to do with the respect you mention, for the work that has gone into some of the issues, and with avoiding a loss of valuable information. Automation is good if it allows us to implement politeness in such a way.

**Comment by brianlheim** [comment]
> The difficulty tag is difficult to assess because it depends on the skills of the person who tackles the bug.

i think we can think of it as the level of effort required for someone who has basic competency in the language or framework at hand; this is how we treat the "good first issue" label -- for instance a C++ "good first issue" isn't an issue that would necessarily be good for someone who's never written C++ before, but it would be good for someone who knows C++ and wants to start on the project. similarly, this tag should help people decide both (1) if they have the knowledge required to fix a bug (combination of this tag and the domain area), and (2) how much time/effort they'll have to spend working on it.

i think "difficulty" is not necessarily the best term, but let's please continue using it for now to avoid bikeshedding.

---

> How to distinguish issues that are closed because they are solved and those closed while still unsolved?

we already do this with labels like "duplicate", "stale", "intended behavior", etc. It would be nice to group those together with the same prefix or color so the distinction is clear. something like "no fix: stale", "no fix: duplicate".

> How to encourage reopening of unsolved stale issues?

i don't want to encourage reopening them, but i do think we should make it clear how and when they can be reopened. we can do that by leaving a boilerplate comment, either manually or with automation.

> How to remind people of the fact that there is an answer pending?

do you mean 'how do we indicate the issue was never implemented/solved'? the "stale" label along with any "blocking:" labels accomplishes that.

> I notice that all these have to do with the respect you mention, for the work that has gone into some of the issues, and with avoiding a loss of valuable information. Automation is good if it allows us to implement politeness in such a way.

closing an issue is definitely a kind of rejection, but the information and work gone into it is not lost or discarded. very often we've had open issues that reference closed issues or PRs as related discussions or 'prior work', and i do this myself when i recognize an issue looks familiar. i agree with the right messaging we can make closing less harsh and avoid misunderstandings.

**Comment by telephon** [comment]
> > How to encourage reopening of unsolved stale issues?
> 
> i don't want to encourage reopening them, but i do think we should make it clear how and when they can be reopened. we can do that by leaving a boilerplate comment, either manually or with automation.

An issue that persists but has been closed due to lack of labor power – shouldn't it be reopened when someone has found a way to tackle it and wants to discuss it?

**Comment by joshpar** [comment]
I’d actually prefer a new issue, with up to date info, be opened.

Trackers, backlogs, etc., are really meant for projects that take a year or so. While I think using them for SC is needed, I would prefer to perhaps think of a way to keep a record of these older issues (for reference), but have a cleaner slate to work off of.

Maybe every major release should have that kind of ‘issue blow-out’?

Josh

**Comment by brianlheim** [comment]
@telephon we're saying the same thing here, just using different terminology. :)

> I’d actually prefer a new issue, with up to date info, be opened.

i would like that. often work on an issue stalls because the discussion gets too long for others to follow, either because it takes off-topic tangents or because it takes a lot of back-and-forth to determine the actual cause of the issue. in the past i've occasionally tried to manage this by updating the top post and hiding irrelevant comments, but it's an uphill battle. in those cases (for me, it's around 10+ comments), having a new issue with a clean summary from the person who is reopening or re-reporting the ticket would be excellent. it would probably make it more likely for the issue to get solved, too. i've seen this happen a few times and it tends to work well.

> Maybe every major release should have that kind of ‘issue blow-out’?

can you elaborate on what you mean by 'issue blow-out'?

**Comment by dyfer** [comment]
> > How to distinguish issues that are closed because they are solved and those closed while still unsolved?
> 
> we already do this with labels like "duplicate", "stale", "intended behavior", etc. It would be nice to group those together with the same prefix or color so the distinction is clear. something like "no fix: stale", "no fix: duplicate".

Yes to a clear label. 
To me a fix that's both a) _valid_ (i.e. not a duplicate, intended behavior, user error etc) and b) _unresolved_ is really a distinctive category. Almost to the point where I'd like to see issues that are "open", "closed" and "unresolved" on top (but that's not possible, is it?). I personally feel like a label is not going to make these stand enough, but that might be the best we could do.

Some automation would be great. I just hope _this_ issue doesn't get closed automatically because of lack of interest haha...

**Comment by brianlheim** [comment]
> Almost to the point where I'd like to see issues that are "open", "closed" and "unresolved" on top (but that's not possible, is it?).

i don't understand what you're describing here.. can you elaborate?

**Comment by dyfer** [comment]
> i don't understand what you're describing here.. can you elaborate?

Sorry if I wasn't clear. I was imagining something like:
![Screen Shot 2020-05-20 at 5 31 27 PM](https://user-images.githubusercontent.com/1589177/82510705-1aaa7b00-9ac0-11ea-8eda-433b0eeda587.png)

Just my imagination!

**Comment by eleses** [comment]
I would honestly caution against closing issues to aggressively. One of the reasons I stopped using Linux distributions from a certain manufacturer is that they closed issues in they bug tracker simply because they hadn't bothered to fix them for a while (even when they were still 100% reproducible in their current distro versions.)

Also long discussions can sometimes be productive (thinking of the one on duplicate cleanups for example.)

**Comment by joshpar** [comment]
I think this is being carefully thought of. We are talking about closing out issues that are years old.

/*
Josh Parmenter
www.realizedsound.net/josh
*/

**Comment by eleses** [comment]
Ok, should this 8+ years old bug report be closed https://github.com/supercollider/supercollider/issues/232

(I just checked, it still happens in 3.11.0.)

I'm pretty sure there are more obscure bug reports like this stuff doesn't work from time to time on that OS... those probably should be closed if nobody can reproduce them. (And I probably filed one or two like that myself.)

**Comment by joshpar** [comment]
I would argue it should be closed, and reopened if it keeps people from doing work. Or, we need a group of people to sift through them, test and verify if they are still an issue and prioritize them. I think that is the spirit of this topic. But given limited dev resources, I would say an issue that is eight years old and hasn’t had a clamor for it being fixed is likely lower priority than a crashing issue or some other recent topics (code signing, delays in server starts in Windows, updating to recent toolchains, etc).

/*
Josh Parmenter
www.realizedsound.net/josh
*/

**Comment by eleses** [comment]
I would suggest then to move the bug tracker off github to some place that allows voting for bugs, if you want to find out how many people care and/or are affected by something. Making users do busy work like reopening old bugs that are still bugs seems pretty counterproductive to me... from a user's perspective.

**Comment by geoffroymontel** [comment]
I am against moving the issues off github because
* even if an issue is closed, we can search for it, and "automatically closed issues" would still be in the issues list, but with a special tag, and a special comment at the end to invite people to reopen a new issue if this bug is important for them
* Github allows people to vote for issues, and sort issues by number of votes, so one option would be to explain this bug cleanup phase to the sc-user mailing list, scsynth forum and slack before doing it, and ask people to vote on issues they still care about.

**Comment by eleses** [comment]
> Github allows people to vote for issues

Do you mean with the [thumb under the comments/posts](https://github.com/Fyrd/caniuse/issues/2408)? I think people don't use that regularly, because they don't presume it's a "keep this bug open, I care about it" vote. (Also there are a bunch of other emoticons in that list, some with no clean meaning.)

**Comment by jamshark70** [comment]
> Github allows people to vote for issues, and sort issues by number of votes, so one option would be to explain this bug cleanup phase to the sc-user mailing list, scsynth forum and slack before doing it, and ask people to vote on issues they still care about.

I think we should establish a protocol for using votes and encourage users to do it.

I used to work in tech support for a business intelligence (it's not actually an oxymoron, though it's tempting to think so) software vendor. The development team had a "pain index," aggregating factors like issue severity, number of support cases logged for that issue, and those customers' support level (the last of those is not relevant for us, of course, because nobody pays for SC support). Engineering effort was a separate measure (so, I can remember one big issue there with a massive pain index, but it went for years without resolution because it needed a huge investment of developer hours).

We have tags for issue severity but essentially no measure at all of users' interest in or need for specific fixes or features. That makes it much harder to prioritize, leading to the situation of hundreds of bug reports and no idea where to start.

It would help to be able to quantify user pain.

**Comment by brianlheim** [comment]
re: voting - be careful with this. not only does it create a discriminatory understanding of user impact -- are you really going to say that accessibility is unimportant just because you see 'only a few people' noting issues with screen readers or a different language / input method? -- but it also does not relate to non-user-facing tickets (such as this one). so, voting alone isn't going to make things easier for maintainers; we still need a separate, visible measure of the severity of an issue.

---

re: moving to another issue tracker. this is a _massive_ risk to take in terms of project longevity and participation. i think we should try to improve on our workflow within GitHub first.

---

i wasn't suggesting to close issues like #232. i don't feel now that setting a project cap or closing issues because they are not 'important' enough is a good idea. but, i do want to close issues where a long discussion has stalled and/or where necessary information for solving the issue isn't available. what do you think of that @eleses ? this sounds different than what you experienced with the Linux distro's bug tracker. if it helps to know this, i am thinking that this stale issue policy would probably amount to closing 2-5 issues per month, if that. perhaps it's wrong to say that this would be more aggressive than what we do now; it's more of a formalization of existing informal practices, in my opinion.

@joshpar , could you accept keeping issues like #232 open even though it is lower-priority in your opinion? i agree with @eleses that closing it and making someone submit a new ticket for it would not be worth the work.

**Comment by eleses** [comment]
@brianlheim No complaints from me for anything you wrote in your last post.

**Comment by brianlheim** [comment]
relevant notes from yesterday's call:

> issue management - https://github.com/supercollider/supercollider/issues/4940
> 
> - Josh P wants to reset everything and start from scratch
> - Luke N also likes this idea, but said we need to have a policy to avoid it happening again
> - Josh P said most projects don't actually run as long as SC. issue trackers work good for shorter term projects, but maybe not for this duration
>   fresh start could give some focus
> - James S explained his ideas regarding stale issues and automation
> - Julian R said a bot would add a kind of neutrality/fairness to it, but is concerned that closing issues will hide valuable information
> - James S suggested updating issue template or other obvious place to note that you may want to search closed issues
> - Everyone agreed messaging around closing issues is important
> - Discussed a gradual approach to transitioning between current process and more controlled approach
> - There was a proposition to have a two-step approach where a bot would comment every issue that’s been stalled for more than X months, which would explain the triage process. Then, after some other time to be determined, the issue would be closed
> - Someone proposed to start with proposition some texts for these messages.

**Comment by adcxyz** [comment]
I can sift through "my" issues as soon as I have a bit more time again. 
Some already had branches with fixes which never got looked at; 
so I'll see if those need updates by now, and make them PRs if possible.

**Comment by dyfer** [comment]
Just to reiterate - I agree that we should choose careful wording and mark old/stale issues.

As for closing them - I wonder if custom search query would be an alternative:
`is:issue is:open -label:stale` will exclude stale issues from the list. Maybe this would be something to consider If there's a way to put a link to this query in a visible place (like where the pinned issues are?) Not sure that's possible though.

**Comment by brianlheim** [comment]
the issues should be closed sometime after going stale. the 'stale' state is just a temporary marker and warning before closing.

**Comment by eleses** [comment]
I think there two kinds of "stale". Stale because it's a clear bug but there's no manpower or interest to fix it. (There a quite a few parts in SC that are basically "we're not doing it this way anymore, but the broken code is still in the code base". And not officially deprecated either.) Or stale because it's debatable if it's even a bug or not easily reproducible etc. 

Perhaps more effort to deprecate and remove the stuff that's not working and there's no interest in fixing it would also help. But that's not alway as solution, e.g. EnvirGui has quite a few little gotchas and bugs when it comes to layout and font changes, but I myself found it not worth it to fix all because I plan to write a new substitute class on better foundation (TreeView). But I'm still *using* EnvirGui in the meantime, with some hacky/quick patches that probably aren't worth the trouble of committing through a formal (code review etc.) process. So it's a perma-purgatory for EnvirGui for me right now, for lack of time to write its "proper" replacement.

**Comment by brianlheim** [comment]
i'm going to close this for now, i think it's served the intended purpose of generating some discussion and ideas. thanks everyone! hoping to come back with a more concrete idea soon.
